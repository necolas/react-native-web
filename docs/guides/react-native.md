# React Native

Use a module loader that supports package aliases (e.g., webpack), and alias
`react-native` to `react-native-web`.

```js
// webpack.config.js

module.exports = {
  // ...
  resolve: {
    alias: {
      'react-native': 'react-native-web'
    }
  }
}
```

## Image assets

In order to require image assets (e.g. `require('assets/myimage.png')`), add an
url-loader to the webpack config:

```js
// webpack.config.js

module.exports = {
  // ...
  module: {
    loaders: {
      // ...
      {test: /\.(jpe?g|png|gif|svg)$/, loader: 'url?name=[path][name].[ext]'}
    }
  }
```

## Web-specific code

Minor platform differences can use the `Platform` module.

```js
import { AppRegistry, Platform, StyleSheet } from 'react-native'

const styles = StyleSheet.create({
  height: (Platform.OS === 'web') ? 200 : 100
})

AppRegistry.registerComponent('MyApp', () => MyApp)

if (Platform.OS === 'web') {
  AppRegistry.runApplication('MyApp', {
    rootTag: document.getElementById('react-root')
  });
}
```

More substantial Web-specific implementation code should be written in files
with the extension `.web.js`, which webpack will automatically resolve.

## Optimizations

Production builds can benefit from dead-code elimination by defining the
following variables:

```js
// webpack.config.js

module.exports = {
  // ...
  plugins: [
    new webpack.DefinePlugin({
      'Platform.OS': 'web',
      'process.env.NODE_ENV': JSON.stringify('production')
    })
  }
}
```
