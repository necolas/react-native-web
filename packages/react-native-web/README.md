# React Native for Web

React Native Components and APIs for the Web.

[Find out more](https://github.com/necolas/react-native-web)
