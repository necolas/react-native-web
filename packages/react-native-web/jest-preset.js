const path = require('path');

module.exports = {
  moduleNameMapper: {
    '^react-native$': path.join(__dirname, 'dist', 'cjs')
  },
  setupFiles: [path.join(__dirname, 'jest', 'setup.js')],
  testEnvironment: require.resolve('jest-environment-jsdom')
};
