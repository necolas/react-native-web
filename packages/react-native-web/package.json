{
  "publishConfig": {
    "registry": "https://registry.npmjs.org/"
  },
  "name": "react-native-web",
  "version": "0.18.9",
  "description": "React Native for Web",
  "module": "dist/index.js",
  "main": "dist/cjs/index.js",
  "sideEffects": false,
  "files": [
    "dist",
    "src",
    "!**/__tests__"
  ],
  "scripts": {
    "clean": "del-cli ./dist",
    "build": "npm-run-all clean -p \"build:* {@}\" --",
    "build:commonjs": "cross-env BABEL_ENV=commonjs babel src --config-file ../../configs/babel.config.js --out-dir dist/cjs --ignore \"**/__tests__\"",
    "build:es": "babel src --config-file ../../configs/babel.config.js --out-dir dist --ignore \"**/__tests__\"",
    "build:flow": "gen-flow-files ./src --out-dir ./dist",
    "dev": "npm-run-all clean -p \"build:* {@} -- --watch\""
  },
  "dependencies": {
    "@babel/runtime": "^7.18.6",
    "create-react-class": "^15.7.0",
    "fbjs": "^3.0.4",
    "inline-style-prefixer": "^6.0.1",
    "normalize-css-color": "^1.0.2",
    "postcss-value-parser": "^4.2.0",
    "styleq": "^0.1.2"
  },
  "peerDependencies": {
    "react": "^17.0.2 || ^18.0.0",
    "react-dom": "^17.0.2 || ^18.0.0"
  },
  "author": "Nicolas Gallagher",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "git://github.com/necolas/react-native-web.git"
  },
  "tags": [
    "react"
  ],
  "keywords": [
    "react",
    "react-component",
    "react-native",
    "web"
  ]
}
