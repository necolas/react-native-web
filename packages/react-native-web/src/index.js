// @flow strict
export { default as unstable_createElement } from './exports/createElement';
export { default as findNodeHandle } from './exports/findNodeHandle';
export { default as processColor } from './exports/processColor';
export { default as render } from './exports/render';
export { default as unmountComponentAtNode } from './exports/unmountComponentAtNode';
export { default as NativeModules } from './exports/NativeModules';

// APIs
export { default as AccessibilityInfo } from './exports/AccessibilityInfo';
export { default as Alert } from './exports/Alert';
export { default as Animated } from './exports/Animated';
export { default as Appearance } from './exports/Appearance';
export { default as AppRegistry } from './exports/AppRegistry';
export { default as AppState } from './exports/AppState';
export { default as BackHandler } from './exports/BackHandler';
export { default as Clipboard } from './exports/Clipboard';
export { default as DeviceInfo } from './exports/DeviceInfo';
export { default as Dimensions } from './exports/Dimensions';
export { default as Easing } from './exports/Easing';
export { default as I18nManager } from './exports/I18nManager';
export { default as Keyboard } from './exports/Keyboard';
export { default as InteractionManager } from './exports/InteractionManager';
export { default as LayoutAnimation } from './exports/LayoutAnimation';
export { default as Linking } from './exports/Linking';
export { default as NativeEventEmitter } from './exports/NativeEventEmitter';
export { default as PanResponder } from './exports/PanResponder';
export { default as PixelRatio } from './exports/PixelRatio';
export { default as Platform } from './exports/Platform';
export { default as Share } from './exports/Share';
export { default as StyleSheet } from './exports/StyleSheet';
export { default as UIManager } from './exports/UIManager';
export { default as Vibration } from './exports/Vibration';

// components
export { default as ActivityIndicator } from './exports/ActivityIndicator';
export { default as Button } from './exports/Button';
export { default as CheckBox } from './exports/CheckBox';
export { default as FlatList } from './exports/FlatList';
export { default as Image } from './exports/Image';
export { default as ImageBackground } from './exports/ImageBackground';
export { default as KeyboardAvoidingView } from './exports/KeyboardAvoidingView';
export { default as Modal } from './exports/Modal';
export { default as Picker } from './exports/Picker';
export { default as Pressable } from './exports/Pressable';
export { default as ProgressBar } from './exports/ProgressBar';
export { default as RefreshControl } from './exports/RefreshControl';
export { default as SafeAreaView } from './exports/SafeAreaView';
export { default as ScrollView } from './exports/ScrollView';
export { default as SectionList } from './exports/SectionList';
export { default as StatusBar } from './exports/StatusBar';
export { default as Switch } from './exports/Switch';
export { default as Text } from './exports/Text';
export { default as TextInput } from './exports/TextInput';
export { default as Touchable } from './exports/Touchable';
export { default as TouchableHighlight } from './exports/TouchableHighlight';
export { default as TouchableNativeFeedback } from './exports/TouchableNativeFeedback';
export { default as TouchableOpacity } from './exports/TouchableOpacity';
export { default as TouchableWithoutFeedback } from './exports/TouchableWithoutFeedback';
export { default as View } from './exports/View';
export { default as VirtualizedList } from './exports/VirtualizedList';
export { default as YellowBox } from './exports/YellowBox';
export { default as LogBox } from './exports/LogBox';

// compat (components)
export { default as DrawerLayoutAndroid } from './exports/DrawerLayoutAndroid';
export { default as InputAccessoryView } from './exports/InputAccessoryView';
export { default as ToastAndroid } from './exports/ToastAndroid';
// compat (apis)
export { default as PermissionsAndroid } from './exports/PermissionsAndroid';
export { default as Settings } from './exports/Settings';
export { default as Systrace } from './exports/Systrace';
export { default as TVEventHandler } from './exports/TVEventHandler';

// plugins
export { default as DeviceEventEmitter } from './exports/DeviceEventEmitter';

// hooks
export { default as useColorScheme } from './exports/useColorScheme';
export { default as useWindowDimensions } from './exports/useWindowDimensions';
