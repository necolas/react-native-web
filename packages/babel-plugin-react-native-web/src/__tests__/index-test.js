const plugin = require('..');
const pluginTester = require('babel-plugin-tester').default;

const tests = [
  // import react-native
  {
    title: 'import from "native-native"',
    code: `import ReactNative from 'react-native';
import { View } from 'react-native';
import { Invalid, View as MyView } from 'react-native';
import * as ReactNativeModules from 'react-native';`,
    snapshot: true
  },
  {
    title: 'import from "native-native"',
    code: `import ReactNative from 'react-native';
import { View } from 'react-native';
import { Invalid, View as MyView } from 'react-native';
import * as ReactNativeModules from 'react-native';`,
    snapshot: true,
    pluginOptions: { commonjs: true }
  },
  {
    title: 'import from "react-native-web"',
    code: `import { unstable_createElement } from 'react-native-web';
import { StyleSheet, View, TouchableOpacity, processColor } from 'react-native-web';
import * as ReactNativeModules from 'react-native-web';`,
    snapshot: true
  },
  {
    title: 'export from "react-native"',
    code: `export { View } from 'react-native';
export { StyleSheet, Text, unstable_createElement } from 'react-native';`,
    snapshot: true
  },
  {
    title: 'export from "react-native-web"',
    code: `export { View } from 'react-native-web';
export { StyleSheet, Text, unstable_createElement } from 'react-native-web';`,
    snapshot: true
  },
  {
    title: 'require "react-native"',
    code: `const ReactNative = require('react-native');
const { View } = require('react-native');
const { StyleSheet, TouchableOpacity } = require('react-native');`,
    snapshot: true
  },
  {
    title: 'require "react-native"',
    code: `const ReactNative = require('react-native');
const { View } = require('react-native');
const { StyleSheet, TouchableOpacity } = require('react-native');`,
    snapshot: true,
    pluginOptions: { commonjs: true }
  },
  {
    title: 'require "react-native-web"',
    code: `const ReactNative = require('react-native-web');
const { unstable_createElement } = require('react-native-web');
const { StyleSheet, View, TouchableOpacity, processColor } = require('react-native-web');`,
    snapshot: true
  }
];

pluginTester({
  babelOptions: {
    generatorOpts: {
      jsescOption: {
        quotes: 'single'
      }
    }
  },
  plugin,
  pluginName: 'Rewrite react-native to react-native-web',
  tests
});
