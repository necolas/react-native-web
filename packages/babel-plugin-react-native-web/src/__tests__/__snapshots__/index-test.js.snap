// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rewrite react-native to react-native-web export from "react-native": export from "react-native" 1`] = `

export { View } from 'react-native';
export { StyleSheet, Text, unstable_createElement } from 'react-native';

      ↓ ↓ ↓ ↓ ↓ ↓

export { default as View } from 'react-native-web/dist/exports/View';
export { default as StyleSheet } from 'react-native-web/dist/exports/StyleSheet';
export { default as Text } from 'react-native-web/dist/exports/Text';
export { default as unstable_createElement } from 'react-native-web/dist/exports/createElement';


`;

exports[`Rewrite react-native to react-native-web export from "react-native-web": export from "react-native-web" 1`] = `

export { View } from 'react-native-web';
export { StyleSheet, Text, unstable_createElement } from 'react-native-web';

      ↓ ↓ ↓ ↓ ↓ ↓

export { default as View } from 'react-native-web/dist/exports/View';
export { default as StyleSheet } from 'react-native-web/dist/exports/StyleSheet';
export { default as Text } from 'react-native-web/dist/exports/Text';
export { default as unstable_createElement } from 'react-native-web/dist/exports/createElement';


`;

exports[`Rewrite react-native to react-native-web import from "native-native": import from "native-native" 1`] = `

import ReactNative from 'react-native';
import { View } from 'react-native';
import { Invalid, View as MyView } from 'react-native';
import * as ReactNativeModules from 'react-native';

      ↓ ↓ ↓ ↓ ↓ ↓

import ReactNative from 'react-native-web/dist/index';
import View from 'react-native-web/dist/exports/View';
import { Invalid } from 'react-native-web/dist/index';
import MyView from 'react-native-web/dist/exports/View';
import * as ReactNativeModules from 'react-native-web/dist/index';


`;

exports[`Rewrite react-native to react-native-web import from "native-native": import from "native-native" 2`] = `

import ReactNative from 'react-native';
import { View } from 'react-native';
import { Invalid, View as MyView } from 'react-native';
import * as ReactNativeModules from 'react-native';

      ↓ ↓ ↓ ↓ ↓ ↓

import ReactNative from 'react-native-web/dist/cjs/index';
import View from 'react-native-web/dist/cjs/exports/View';
import { Invalid } from 'react-native-web/dist/cjs/index';
import MyView from 'react-native-web/dist/cjs/exports/View';
import * as ReactNativeModules from 'react-native-web/dist/cjs/index';


`;

exports[`Rewrite react-native to react-native-web import from "react-native-web": import from "react-native-web" 1`] = `

import { unstable_createElement } from 'react-native-web';
import { StyleSheet, View, TouchableOpacity, processColor } from 'react-native-web';
import * as ReactNativeModules from 'react-native-web';

      ↓ ↓ ↓ ↓ ↓ ↓

import unstable_createElement from 'react-native-web/dist/exports/createElement';
import StyleSheet from 'react-native-web/dist/exports/StyleSheet';
import View from 'react-native-web/dist/exports/View';
import TouchableOpacity from 'react-native-web/dist/exports/TouchableOpacity';
import processColor from 'react-native-web/dist/exports/processColor';
import * as ReactNativeModules from 'react-native-web/dist/index';


`;

exports[`Rewrite react-native to react-native-web require "react-native": require "react-native" 1`] = `

const ReactNative = require('react-native');
const { View } = require('react-native');
const { StyleSheet, TouchableOpacity } = require('react-native');

      ↓ ↓ ↓ ↓ ↓ ↓

const ReactNative = require('react-native-web/dist/index');

const View = require('react-native-web/dist/exports/View').default;

const StyleSheet = require('react-native-web/dist/exports/StyleSheet').default;

const TouchableOpacity = require('react-native-web/dist/exports/TouchableOpacity').default;


`;

exports[`Rewrite react-native to react-native-web require "react-native": require "react-native" 2`] = `

const ReactNative = require('react-native');
const { View } = require('react-native');
const { StyleSheet, TouchableOpacity } = require('react-native');

      ↓ ↓ ↓ ↓ ↓ ↓

const ReactNative = require('react-native-web/dist/cjs/index');

const View = require('react-native-web/dist/cjs/exports/View').default;

const StyleSheet = require('react-native-web/dist/cjs/exports/StyleSheet').default;

const TouchableOpacity = require('react-native-web/dist/cjs/exports/TouchableOpacity').default;


`;

exports[`Rewrite react-native to react-native-web require "react-native-web": require "react-native-web" 1`] = `

const ReactNative = require('react-native-web');
const { unstable_createElement } = require('react-native-web');
const { StyleSheet, View, TouchableOpacity, processColor } = require('react-native-web');

      ↓ ↓ ↓ ↓ ↓ ↓

const ReactNative = require('react-native-web/dist/index');

const unstable_createElement = require('react-native-web/dist/exports/createElement').default;

const StyleSheet = require('react-native-web/dist/exports/StyleSheet').default;

const View = require('react-native-web/dist/exports/View').default;

const TouchableOpacity = require('react-native-web/dist/exports/TouchableOpacity').default;

const processColor = require('react-native-web/dist/exports/processColor').default;


`;
